### PR TITLE
[SPARK-45556][UI] Allow web page respond customized status code and message through WebApplicationException

### PR DESCRIPTION
**Pull request by kuwii**
### What changes were proposed in this pull request?

Adds an extra catch of `WebApplicationException` in Spark core and history server's web page servlet, to let customized response to pass-through instead of wrapping them in a 500 response, to make web page and REST API have consistent behavior.

After the change, when a `WebApplicationException` is thrown, servlet will render a basic Spark page with the exception message, and respond with the status code of the exception.

For example, previously when a customized 418 exception is thrown, the response will be a 500 like this:
image

After the change, a 418 response will be rendered like this:
after

### Why are the changes needed?

Currently for web page servlet in Spark, when an exception is thrown, there's only 3 possible responses:
- If it is history server unable to find an app, a 404 will be responded.
- It the exception is an `IllegalArgumentException`, a 400 will be responded.
- Otherwise, a 500 will be responded, wrapping the original exception within it.

However for REST API, if a response is a `WebApplicationException`, its response can be directly returned with the designed status code and message, without being wrapped. This causes inconsistent behavior between web page and REST API.

For history server, Spark allows to use alternative `ApplicationHistoryProvider`, or adds extra page and API through "AppHistoryServerPlugin". Suppose some extra provider or plugins is used, and in some cases they want to return a special status code due to their needs. Currently there's no way to do it other than forking Spark because web page servlet does not support it.

### Does this PR introduce _any_ user-facing change?

By default, no. Because Spark UI's built-in web page servlet does not throw `WebApplicationException`.

For 3rd party provider or plugin, if a `WebApplicationException` is thrown, after this change, its message and status code can be directly responded. Previously it is wrapped within a 500 response.

### How was this patch tested?

- 3 UTs are added:
  - Testing whether history server servlet can handle `WebApplicationException`
  - Testing whether servlet created by `JettyUtils` can handle `WebApplicationException`
  - Testing whether servlet created by `JettyUtils` can handle other exceptions as before
- Manually tested

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Would be great if we can contain the screenshot of UI with this fix.

**Comment by kuwii** [comment]
> Would be great if we can contain the screenshot of UI with this fix.

Thanks for the comment. Have updated the screenshot before and after the change.

**Comment by kuwii** [comment]
Kindly ping @HyukjinKwon . Could you please help to take a look at this PR if you think it is OK to do this change? Thanks.

**Comment by kuwii** [comment]
Also kindly ping @srowen @dongjoon-hyun.

File: core/src/test/scala/org/apache/spark/deploy/history/HistoryServerSuite.scala [hunk 1]
```diff
@@ -93,16 +95,23 @@ abstract class HistoryServerSuite extends SparkFunSuite with BeforeAndAfter with
       .set(EXECUTOR_PROCESS_TREE_METRICS_ENABLED, true)
       .set(HYBRID_STORE_DISK_BACKEND, diskBackend.toString)
     conf.setAll(extraConf)
-    provider = new FsHistoryProvider(conf)
-    provider.checkForLogs()
+    val usedProvider = appHistoryProvider.getOrElse {
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/deploy/history/HistoryServerSuite.scala:98]: Is this a related change?

**Review by srowen — COMMENTED** [review]
Seems OK. It's a little bit better than the raw stack trace. I don't suppose there's a similar change to make in the main Spark UI? or does it also already work this way?

File: core/src/test/scala/org/apache/spark/deploy/history/HistoryServerSuite.scala (hunk 1 above)
Comment (kuwii) [reply core/src/test/scala/org/apache/spark/deploy/history/HistoryServerSuite.scala:98]: Yes, to test whether history server return correct response, the UT added in this file requires to create history server service with a private implementation of provider interface, to throw customized error. But current version only creates `FsHistoryProvider`.

So change here is to allow to create pass private provider to initialization method. The previous init way is still kept, so it won't affect other existing tests.

image

image

**Comment by kuwii** [comment]
> Seems OK. It's a little bit better than the raw stack trace. I don't suppose there's a similar change to make in the main Spark UI? or does it also already work this way?

I don't think Spark's built-in UI needs this change, because web page doesn't seem to throw `WebApplicationException`, and REST API already supports this. The gap may be because web page and REST API are using different ways to handle HTTP requests.

The main goal of this PR is to let web page part also able to return customized response code (like 418 as an example), instead of just 200, 400, 404 and 500. It is basically for those who wants to extend Spark UI through plugins, without forking Spark source code.

**Comment by kuwii** [comment]
Kindly ping @srowen
